### PR TITLE
Sync reordering

### PR DESCRIPTION
**Pull request by iccub**
Notes:
- copied bookmark_util.js from crypto lib to a new file, to utilize synchronous javascript calls. It made reordering work easier to do.
- sync reordering algorithms is now used even if not in a sync group. The exception is favorites, where old `order` attribute is being used.
- removing `syncUUID` while joining a sync group is not implemented yet
- there is an update bug, I want to fix it on a separate branch(it was introduced before my changes)

### CODE REVIEW

File: brave/src/sync/Sync.swift [hunk 1]
```diff
@@ -570,9 +609,32 @@ extension Sync {
...
 
+    }
+    
+    func getBookmarkOrder(previousOrder: String?, nextOrder: String?) -> String? {
+        
+        // Empty string as a parameter means next/previous bookmark doesn't exist
+        let prev = previousOrder ?? ""
+        let next = nextOrder ?? ""
+        
+        let function = jsContext?.objectForKeyedSubscript("getBookmarkOrder")
```
Comment (jhreis) [inline brave/src/sync/Sync.swift]: `function` should be more descriptive.

File: brave/src/sync/Sync.swift [hunk 2]
```diff
@@ -570,9 +609,32 @@ extension Sync {
...
+        return function?.call(withArguments: [prev, next]).toString()
+    }
+    
+    fileprivate func setBaseBookmarkOrder() {
+        guard let deviceId = Device.currentDevice()?.deviceId?.first else { return }
+        let function = jsContext?.objectForKeyedSubscript("getBaseBookmarksOrder")
```
Comment (jhreis) [inline brave/src/sync/Sync.swift]: `function` should be more descriptive.

File: brave/src/sync/Sync.swift [hunk 3]
```diff
@@ -355,6 +389,11 @@ extension Sync {
             return
         }
         
+        if let syncBookmark = records.first?.recordType, syncBookmark == .bookmark, baseSyncOrder == nil {
```
Comment (jhreis) [inline brave/src/sync/Sync.swift]: Since equality can just do:

`records.first?.recordType == .bookmark && baseSyncOrder == nil {}`

Comment (jhreis) [inline brave/src/sync/Sync.swift]: The check here can be removed, as this actually automatically handles this: https://github.com/brave/browser-ios/pull/1636/files#diff-d4a71e14f4d1211c6b1ea257fdbe5e94R382

File: brave/src/sync/Sync.swift [hunk 4]
```diff
@@ -121,6 +132,29 @@ class Sync: JSInjector {
...
+        context?.exceptionHandler = { _, exc in
+            log.error(exc.debugDescription)
+        }
+        guard let path = Bundle.main.path(forResource: "bookmark_util", ofType: "js") else {
+            log.error("Could not load bookmark_util.js")
+            return nil
+        }
+        
+        do {
+            let scriptAsString = try String(contentsOfFile: path, encoding: String.Encoding.utf8)
```
Comment (jhreis) [inline brave/src/sync/Sync.swift]: Leave some comments in here if possible, a bit weird seeing the JS loaded into a context, and result ignored (I know it is to leverage the usage later).

File: brave/src/sync/Sync.swift [hunk 5]
```diff
@@ -608,6 +671,7 @@ extension Sync: WKScriptMessageHandler {
             print("---- Sync Debug: \(data)")
         case "sync-ready":
             self.isSyncFullyInitialized.syncReady = true
+            setBaseBookmarkOrder()
```
Comment (jhreis) [inline brave/src/sync/Sync.swift]: Any reason this isn't inside `checkIsSyncReady`?

File: brave/src/sync/Sync.swift [hunk 6]
```diff
@@ -599,6 +661,7 @@ extension Sync: WKScriptMessageHandler {
             // (e.g. arg2 is [Int])
             let data = JSON(parseJSON: message.body as? String ?? "")
             self.saveInitData(data)
+            Bookmark.removeSyncOrders()
```
Comment (jhreis) [inline brave/src/sync/Sync.swift]: Not understanding this one, can you explain?

File: brave/src/data/Bookmark.swift [hunk 7]
```diff
@@ -257,30 +282,62 @@ public final class Bookmark: NSManagedObject, WebsitePresentable, Syncable, CRUD
...
-            bms.insert(src, at: bms.index(of: dest)!)
-        } else {
-            let end = bms.index(of: dest)! + 1
-            bms.insert(src, at: end)
-        }
-        
-        for i in 0..<bms.count {
-            bms[i].order = Int16(i)
+        // Favorites use regular order algorithm
+        if src.isFavorite {
```
Comment (jhreis) [inline brave/src/data/Bookmark.swift]: These need to be split up, this function is massive.

File: brave/src/sync/Sync.swift (hunk 6 above)
Comment (iccub) [reply brave/src/sync/Sync.swift]: `syncOrder` is now used internally for ordering even if user is not using sync. So bookmarks go like `0.0.1, 0.0.2...` after we join a sync group we are removing these sync order and setting them again, now with a proper device id, so we can replace `0.x.x` with `device_id.x.x`

Comment (jhreis) [reply brave/src/sync/Sync.swift]: But this will be called every time sync is initialized, right?

Comment (iccub) [reply brave/src/sync/Sync.swift]: Yes, but sync is initialized only once. Is it a problem?
New sync join = new device id, we need to update bookmarks anyway

Comment (jhreis) [reply brave/src/sync/Sync.swift]: Ah, I thought it was called every time the library was initialized, as in, every launch. Would add comment here, but lgtm.

File: brave/src/data/Bookmark.swift [hunk 8]
```diff
@@ -219,7 +230,21 @@ public final class Bookmark: NSManagedObject, WebsitePresentable, Syncable, CRUD
         
         return bk
     }
-  
+    
+    class func allBookmarksOfAGivenLevelPredicate(parent: Bookmark?) -> NSPredicate {
+        let isFavoriteKP = #keyPath(Bookmark.isFavorite)
+        let parentFolderKP = #keyPath(Bookmark.parentFolder)
+        
+        if let parentFolder = parent {
```
Comment (jhreis) [inline brave/src/data/Bookmark.swift]: Can this be reduced to:
`return NSPredicate(format: "%K == %@ AND %K == NO", parentFolderKP, parent ?? "nil", isFavoriteKP)`

File: brave/src/data/Bookmark.swift [hunk 9]
```diff
@@ -257,30 +282,62 @@ public final class Bookmark: NSManagedObject, WebsitePresentable, Syncable, CRUD
...
+            let isMovingUp = sourceIndexPath.row > destinationIndexPath.row
+            
+            // Depending on drag direction, all other bookmarks are pushed up or down. 
+            if isMovingUp {
+                var prev: String?
+                
+                // Bookmark at the top has no previous bookmark.
+                if destinationIndexPath.row > 0 {
+                    let index = IndexPath(row: destinationIndexPath.row - 1, section: destinationIndexPath.section)
+                    prev = (frc.object(at: index) as! Bookmark).syncOrder
```
Comment (jhreis) [inline brave/src/data/Bookmark.swift]: We should remove all of these forced castings. Change https://github.com/brave/browser-ios/pull/1636/files#diff-3ec30f0d026baec8beed697836bb1b02R276

to:
`guard let frc = frc as? NSFetchedResultsController<Bookmark> else { return }`

This allows `frc` to always return `Bookmark`

File: brave/src/sync/Sync.swift (hunk 5 above)
Comment (iccub) [reply brave/src/sync/Sync.swift]: This is causing race conditions as well, going to move it to `checkIsSyncReady`.

File: brave/src/sync/Sync.swift (hunk 4 above)
Comment (iccub) [reply brave/src/sync/Sync.swift]: I added a comment one line below :) just fyi bc github didn't collapse this comment.

File: brave/src/sync/bookmark_util.js [hunk 10]
```diff
@@ -0,0 +1,125 @@
+'use strict'
+
+
```
Comment (iccub) [inline brave/src/sync/bookmark_util.js]: I'm thinking about adding a comment here, to not forget that this is our handmade converted file from `brave/sync`

File: brave/src/sync/Sync.swift [hunk 11]
```diff
@@ -355,6 +410,12 @@ extension Sync {
             return
         }
         
+        if recordType == .bookmark && baseSyncOrder == nil {
```
[the remark this answers is not in the document]
Comment (iccub) [reply brave/src/sync/Sync.swift]: I'm not sure if I get it.

These are separate checks,
first check unwraps record type, and second check only happens if record type is a bookmark.
Recordtype can be a device, preferences etc. so this second check is only for sync record types.

For example if we are sending `.device` type of record type, we do not care about `baseSyncOrder`

**Pull request merged**